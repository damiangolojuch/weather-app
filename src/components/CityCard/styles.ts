import styled from "styled-components";
import { Card } from "antd";
import WeatherIcon from "~/components/WeatherIcon";

export const CardStyled = styled(Card)`
  border-radius: 10px;
  background-color: rgba(255, 255, 255, 0.9);
`;

export const WeatherIconWrapper = styled.div`
  text-align: center;
  width: 100%;
  position: relative;
`;

export const WeatherIconStyled = styled(WeatherIcon)`
  width: 100%;
  height: 100%;
  max-height: 25vh;
  max-width: 25vh;
`;
