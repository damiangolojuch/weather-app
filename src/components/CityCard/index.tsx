import { FC } from "react";
import * as S from "./styles";
import { CityWeatherType } from "~/modules/WeatherLogic/constants";
import { Link } from "react-router-dom";
import CityStatistic from "~/components/CityStatistic";
import { CityStatisticType } from "~/components/CityStatistic/constants";
import { WeatherIcons } from "~/components/WeatherIcon/constants";

export interface CityCardProps {
  cityData: CityWeatherType;
  compareToCityData?: CityWeatherType;
}

const statistics = [
  CityStatisticType.windSpeed,
  CityStatisticType.temperature,
  CityStatisticType.pressure,
  CityStatisticType.humidity,
];

const CityCard: FC<CityCardProps> = ({ cityData, compareToCityData }) => (
  <Link to={`/${cityData.city}`}>
    <S.CardStyled hoverable title={cityData.city}>
      <S.WeatherIconWrapper>
        <S.WeatherIconStyled
          type={cityData.clouds === 0 ? WeatherIcons.day : WeatherIcons.rainy6}
        />
      </S.WeatherIconWrapper>
      {statistics.map((statistic) => (
        <CityStatistic
          key={statistic}
          cityData={cityData}
          compareToCityData={compareToCityData}
          type={statistic}
        />
      ))}
    </S.CardStyled>
  </Link>
);

export default CityCard;
