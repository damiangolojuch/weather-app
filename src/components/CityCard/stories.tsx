import { ComponentStory, ComponentMeta } from "@storybook/react";
import CityCard from ".";

export default {
  title: "CityCard",
  component: CityCard,
  args: {
    description: "Displays city statistics or compare to other city.",
  },
} as ComponentMeta<typeof CityCard>;

const DefaultTemplate: ComponentStory<typeof CityCard> = (args) => (
  <CityCard {...args} />
);

export const Default = DefaultTemplate.bind({});

Default.args = {
  cityData: {
    city: "Kraków",
    id: "123123",
    clouds: 0,
    windSpeed: 12,
    temperature: 31,
    pressure: 1002,
    humidity: 51,
  },
};

const CompareTemplate: ComponentStory<typeof CityCard> = (args) => (
  <CityCard {...args} />
);

export const Compare = CompareTemplate.bind({});

Compare.args = {
  cityData: {
    city: "Kraków",
    id: "123123",
    clouds: 0,
    windSpeed: 12,
    temperature: 31,
    pressure: 1002,
    humidity: 51,
  },
  compareToCityData: {
    city: "Warszawa",
    id: "123125",
    clouds: 1,
    windSpeed: 10,
    temperature: 28,
    pressure: 1012,
    humidity: 31,
  },
};
