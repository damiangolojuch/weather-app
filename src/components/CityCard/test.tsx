import CityCard, { CityCardProps } from ".";
import { screen } from "@testing-library/react";
import renderWithRouter from "~/utils/tests/renderWithRouter";

jest.mock("components/CityStatistic", () => {
  return {
    __esModule: true,
    default: (props: any) => (
      <div className="CityStatistic" id={props.type}>
        {props.children}
      </div>
    ),
  };
});

jest.mock("components/WeatherIcon", () => {
  return {
    __esModule: true,
    default: (props: any) => (
      <div className="WeatherIcon">{props.children}</div>
    ),
  };
});

jest.mock("antd", () => {
  const ant = jest.requireActual("antd");
  return {
    __esModule: true,
    ...ant,
    Card: (props: any) => <div className="Card">{props.children}</div>,
  };
});

const setup = (customProps: Partial<CityCardProps> = {}) => {
  const defaultProps: CityCardProps = {
    cityData: {
      city: "SomeCity",
      humidity: 12,
      pressure: 1001,
      temperature: 31,
      windSpeed: 9,
      clouds: 1,
      id: "123123",
    },
    ...customProps,
  };

  return renderWithRouter(<CityCard {...defaultProps} />);
};

describe("<CityCard />", () => {
  it("should contain link to passed city", () => {
    const { container } = setup();

    expect(screen.getByRole("link")).toBeInTheDocument();
    expect(container.getElementsByTagName("a").item(0)).toHaveAttribute(
      "href",
      "/SomeCity"
    );
  });

  it("should render ant design's Card component with WeatherIcon and Statistics component", () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it("should render Statistics component for windSpeed, temperature, pressure and humidity", () => {
    const { container } = setup();

    expect(
      container.getElementsByClassName("CityStatistic").item(0)
    ).toHaveAttribute("id", "windSpeed");
    expect(
      container.getElementsByClassName("CityStatistic").item(1)
    ).toHaveAttribute("id", "temperature");
    expect(
      container.getElementsByClassName("CityStatistic").item(2)
    ).toHaveAttribute("id", "pressure");
    expect(
      container.getElementsByClassName("CityStatistic").item(3)
    ).toHaveAttribute("id", "humidity");
  });
});
