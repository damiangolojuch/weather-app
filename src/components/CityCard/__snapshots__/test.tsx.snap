// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CityCard /> should render ant design's Card component with WeatherIcon and Statistics component 1`] = `
<div>
  <a
    href="/SomeCity"
  >
    <div
      class="Card"
    >
      <div
        class="sc-gsnTZi lnZKao"
      >
        <div
          class="WeatherIcon"
        />
      </div>
      <div
        class="CityStatistic"
        id="windSpeed"
      />
      <div
        class="CityStatistic"
        id="temperature"
      />
      <div
        class="CityStatistic"
        id="pressure"
      />
      <div
        class="CityStatistic"
        id="humidity"
      />
    </div>
  </a>
</div>
`;
